demos/guide/quic-server-block.c

```diff
@@ -0,0 +1,312 @@
+    SSL_CTX_set_alpn_select_cb(ctx, select_alpn, NULL);
+
+    return ctx;
+
+err:
+    SSL_CTX_free(ctx);
+    return NULL;
+}
+
+/* Create UDP socket on the given port. */
+static int create_socket(uint16_t port)
+{
+    int fd;
+    struct sockaddr_in sa = {0};
+
+    /* Retrieve the file descriptor for a new UDP socket */
+    if ((fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0) {
+        fprintf(stderr, "cannot create socket");
+        goto err;
+    }
+
+    sa.sin_family = AF_INET;
+    sa.sin_port = htons(port);
+
+    /* Bind to the new UDP socket on localhost */
+    if (bind(fd, (const struct sockaddr *)&sa, sizeof(sa)) < 0) {
+        fprintf(stderr, "cannot bind to %u\n", port);
+        BIO_closesocket(fd);
+        goto err;
+    }
+
+    return fd;
+
+err:
+    BIO_closesocket(fd);
+    return -1;
+}
+
+/*
+ * Main loop for server to accept QUIC connections.
+ * Echo every request back to the client.
+ */
+static int run_quic_server(SSL_CTX *ctx, int fd)
+{
+    int ok = 0;
+    SSL *listener, *conn;
+    unsigned char buf[8192];
+    size_t nread;
+    size_t nwritten;
+
+    /*
+     * Create a new QUIC listener. Listeners, and other QUIC objects, default
+     * to operating in blocking mode. The configured behaviour is inherited by
+     * child objects.
+     */
+    if ((listener = SSL_new_listener(ctx, 0)) == NULL) {
+        goto err;
+    }
+
+    /* Provide the listener with our UDP socket. */
+    if (!SSL_set_fd(listener, fd))
+        goto err;
+
+    /* Begin listening. */
+    if (!SSL_listen(listener))
+        goto err;
+
+    /*
+     * Begin an infinite loop of listening for connections. We will only
+     * exit this loop if we encounter an error.
+     */
+    for (;;) {
+        /* Pristine error stack for each new connection */
+        ERR_clear_error();
+
+        /* Block while waiting for a client connection */
+        printf("Waiting for connection\n");
+        conn = SSL_accept_connection(listener, 0);
+        if (conn == NULL) {
+            fprintf(stderr, "error while accepting connection\n");
+            goto err;
+        }
+        printf("Accepted new connection\n");
+
+        /* Echo client input */
+        while (SSL_read_ex(conn, buf, sizeof(buf), &nread) > 0) {
+            if (SSL_write_ex(conn, buf, nread, &nwritten) > 0 &&
+                nwritten == nread) {
+                break;
+            }
+            fprintf(stderr, "Error echoing client input");
+            break;
+        }
+
+        /* Signal the end of the stream. */
+        if (SSL_stream_conclude(conn, 0) != 1) {
+            fprintf(stderr, "Unable to conclude stream\n");
+            SSL_free(conn);
+            goto err;
+        }
+
+        /*
+         * Shut down the connection. We may need to call this multiple times
+         * to ensure the connection is shutdown completely.
+         */
+        SSL_SHUTDOWN_EX_ARGS shutdown_args = {0};
```

Comment:
Style nit: mixed declarations/code isn't c90. Put this at the beginning of the block